File: client/src/services/advisor.ts
import { CONFIG } from "../config";

export type AdvisorChatRole = "user" | "assistant";

export interface AdvisorChatMessage {
  role: AdvisorChatRole;
  content: string;
}

export interface AdvisorChatContext {
  active_tab?: string;
  search_query?: string;
  product_id?: string;
}

export interface AdvisorPlatformRecommendation {
  platform: string;
  price: number | null;
  url: string | null;
  in_stock: boolean | null;
}

export interface AdvisorRecommendation {
  product_id: string;
  product_name: string;
  reason: string;
  lowest_price: number | null;
  platforms: AdvisorPlatformRecommendation[];
}

export interface AdvisorSource {
  type: string;
  id: string;
  label: string;
}

export interface AdvisorChatResponse {
  answer: string;
  recommendations: AdvisorRecommendation[];
  sources: AdvisorSource[];
}

export async function sendAdvisorMessage(
  message: string,
  history: AdvisorChatMessage[],
  context: AdvisorChatContext,
): Promise<AdvisorChatResponse> {
  const response = await fetch(`${CONFIG.API_URL}/advisor/chat`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message,
      history,
      context,
    }),
  });

  if (!response.ok) {
    const detail = await response.json().catch(() => null);
    throw new Error(detail?.detail || "Không thể kết nối ProductHunter Advisor");
  }

  return response.json();
}

File: server/app/schemas/advisor.py
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


ChatRole = Literal["user", "assistant"]


class AdvisorChatMessage(BaseModel):
    role: ChatRole
    content: str = Field(..., min_length=1, max_length=4000)


class AdvisorChatContext(BaseModel):
    active_tab: Optional[str] = None
    search_query: Optional[str] = None
    product_id: Optional[UUID] = None


class AdvisorChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=4000)
    history: list[AdvisorChatMessage] = Field(default_factory=list, max_length=12)
    context: Optional[AdvisorChatContext] = None


class AdvisorPlatformRecommendation(BaseModel):
    platform: str
    price: Optional[float] = None
    url: Optional[str] = None
    in_stock: Optional[bool] = None


class AdvisorRecommendation(BaseModel):
    product_id: UUID
    product_name: str
    reason: str
    lowest_price: Optional[float] = None
    platforms: list[AdvisorPlatformRecommendation] = Field(default_factory=list)


class AdvisorSource(BaseModel):
    type: str
    id: str
    label: str


class AdvisorChatResponse(BaseModel):
    answer: str
    recommendations: list[AdvisorRecommendation] = Field(default_factory=list)
    sources: list[AdvisorSource] = Field(default_factory=list)
